hack/generator/pkg/astmodel/strip_unused.go

```diff
@@ -0,0 +1,177 @@
+func StripUnusedDefinitions(
+	roots TypeNameSet,
+	definitions []TypeDefiner,
+) ([]TypeDefiner, error) {
+	// Build a referrers map for each type.
+	referrers := make(map[TypeName]TypeNameSet)
+
+	for _, def := range definitions {
+		for reference := range def.Type().References() {
+			name := def.Name()
+			if name == nil {
+				klog.V(5).Infof("nil name for %#v", def)
+			}
+			referrers[reference] = referrers[reference].Add(*def.Name())
+		}
+	}
+
+	klog.V(0).Infof("definitions: %d, roots: %d, referrers: %d", len(definitions), len(roots), len(referrers))
+
+	checker := newConnectionChecker(roots, referrers)
+	var newDefinitions []TypeDefiner
+	for _, def := range definitions {
+		if def.Name() == nil {
+			klog.V(5).Infof("nil name for definition %#v", def)
+			continue
+		}
+		if checker.connected(*def.Name()) {
+			newDefinitions = append(newDefinitions, def)
+		}
+	}
+	return newDefinitions, nil
+}
+
+// CollectResourceDefinitions returns a TypeNameSet of all of the
+// resource definitions in the definitions passed in.
+func CollectResourceDefinitions(definitions []TypeDefiner) TypeNameSet {
+	resources := make(TypeNameSet)
+	for _, def := range definitions {
+		if _, ok := def.(*ResourceDefinition); ok {
+			if def.Name() == nil {
+				klog.V(5).Infof("nil name for %#v", def)
+			}
+			resources.Add(*def.Name())
+		}
+	}
+	return resources
+}
+
+func newConnectionChecker(roots TypeNameSet, referrers map[TypeName]TypeNameSet) *connectionChecker {
+	return &connectionChecker{
+		roots:     roots,
+		referrers: referrers,
+	}
+}
+
+type connectionChecker struct {
+	roots     TypeNameSet
+	referrers map[TypeName]TypeNameSet
+
+	// memo tracks results for typenames we've already seen.
+	memo TypeNameSet
+}
+
+func (c *connectionChecker) connected(name TypeName) bool {
+	return c.checkWithPath(name, nil)
+}
+
+func (c *connectionChecker) checkWithPath(name TypeName, path TypeNameSet) bool {
+	if c.roots.Contains(name) {
+		return true
+	}
+	// We don't need to recheck for a type we've seen is connected.
+	if c.memo.Contains(name) {
+		return true
+	}
+	path = path.Add(name)
+	for referrer := range c.referrers[name] {
+		if path.Contains(referrer) {
+			// We've already visited this type, don't get caught
+			// in a cycle.
+			continue
+		}
+		if c.checkWithPath(referrer, path) {
+			// If our referrer is connected to a root, then we are
+			// too - track that in memo. Parent callers will
+			// record for themselves, so we don't need to store
+			// all of path
+			c.memo = c.memo.Add(name)
+			return true
+		}
+	}
+	path.Remove(name)
+
+	// Note: We can't memoise the negative result here because we
+	// don't know whether the search failed because it would have
+	// included a cycle. It's possible for this call to fail but a
+	// search taking a different path to this node still to succeed
+	// and prove that this is connected to a root. (It took me longer
+	// than I'd like to understand this.)
+	return false
```

Comment:
I'm not sure I follow this either -- might sync up with you offline on it.

Reply:
Goes away with your nice simpler way.